plugins/Jobs.cpp

```diff
@@ -1232,7 +1232,7 @@ bool BedrockPlugin_Jobs::processCommand(SQLite& db, BedrockCommand& command) {
         if (jobIDs.size()) {
             const string& name = request["name"];
             string nameQuery = name.empty() ? "" : ", name = " + SQ(name) + "";
-            string updateQuery = "UPDATE jobs SET state = 'QUEUED', nextRun = DATETIME("+ SCURRENT_TIMESTAMP() + ")"+ nameQuery +" WHERE jobID IN(" + SQList(jobIDs)+ ");";
+            string updateQuery = "UPDATE jobs SET state = 'QUEUED', nextRun = created"+ nameQuery +" WHERE jobID IN(" + SQList(jobIDs)+ ");";
```

Comment:
Why are you updating it to created? I think you should leave it unchanged.

Reply:
I dropped context for this [here in slack](https://expensify.slack.com/archives/C03TQ48KC/p1573598379133700?thread_ts=1573571868.102800&cid=C03TQ48KC)

Reply:
From that message:
> Circling back to this, we cannot easily just remove the update to nextRun. When we fetch a chat for an agent, we set nextRun to be ~2 hours in the future. This is because GetJob(s) also returns jobs in a RUNQUEUED state, so we could potentially double assign a chat if we did not update nextRun.

I don't get the:
> we could potentially double assign a chat if we did not update nextRun.

How is that?

Reply:
> I don't get the:
> 
> > we could potentially double assign a chat if we did not update nextRun.
> 
> How is that?

Since `GetJob(s)` fetches jobs in _either_ `QUEUED` or `RUNQUEUED` states. We want to get jobs in both because it ensures that after an unclean exit (killing the tab), these chat's will not permanently be lost. Since we set `nextRun` [to the future of all `ChatbotProcessInputAsync` jobs](https://github.com/Expensify/Bedrock/blob/master/plugins/Jobs.cpp#L805) when we fetch them, they will naturally requeue themselves after the `nextRun` time is hit. This also ensures that the jobs are not double fetched in this time.

If we leave `nextRun` as is here, we run into a similar, almost worse, problem where exiting a chat will basically freeze it for ~2 hours, as well as pushing it to the back of the line again.

Reply:
ok I understand part of it (I think). Since some jobs have a `retryAfter`, when we get them the first time with GetJob(s), it sets the next run based on the `retryAfter`, so when we requeue if we don't modify the next run, it will not be actually be dequeued till the retryAfter time passes. Correct? If so, we should reverse the nextRunAt (by doing the inverse logic we do to increase the nextRun based on the retryAfter) if the job has a retryAfter set. 

> This also ensures that the jobs are not double fetched in this time.

I still don't understand this, AFAIK there's no way of processing the same job twice (unless the retryAfter has passed).

Reply:
> ok I understand part of it (I think). Since some jobs have a `retryAfter`, when we get them the first time with GetJob(s), it sets the next run based on the `retryAfter`, so when we requeue if we don't modify the next run, it will not be actually be dequeued till the retryAfter time passes. Correct? If so, we should reverse the nextRunAt (by doing the inverse logic we do to increase the nextRun based on the retryAfter) if the job has a retryAfter set.

I like this solution better, will change to this.

> > This also ensures that the jobs are not double fetched in this time.
> 
> I still don't understand this, AFAIK there's no way of processing the same job twice (unless the retryAfter has passed).

We don't actually ever set the state to `RUNNING` for `ChatbotProcessInputAsync`. We only ever fetch the job, which will move the job to be in the `RUNQUEUED` state. Since `GetJob` will return jobs in either 
`QUEUED` or `RUNQUEUED`, recalling the job _without_ updating the `nextRun` time will result in a chat that has already been fetched being fetched again.

Reply:
> If so, we should reverse the nextRunAt (by doing the inverse logic we do to increase the nextRun based on the retryAfter) if the job has a retryAfter set.

So after thinking about this and working on it all day, I don't think we can actually do this easily. We generate a new `nextRun` time based on the current timestamp when we get the job. We don't have a way to get back to the original `nextRun` time short of storing that time and resetting to that. I don't see huge benefits to doing that over simply setting the `nextRun` time to the created. [Here is where we do the `nextRun` updating.](https://github.com/Expensify/Bedrock/blob/609594240ee58a6d89f3e56580c7490f786a7fb3/plugins/Jobs.cpp#L807-L810)

Do you have a suggested way around that?

Reply:
hmmmm, isn't the value we want the nextRun minus the retry after?

Reply:
> hmmmm, isn't the value we want the nextRun minus the retry after?
Here's an example as to why that won't work, and how it currently behaves, how your proposed solution would work, and how my solution would work.

For this example, we will assume 1 chat is fetched, not 3, for sake of simplicity. I will also be excluding any dates, since all of this will be on the same date, and any columns that do not affect this. 
# Initial State
1. User 1 sends in a message at 04:00. This is jobID 1
2. User 2 sends in a message at 05:00. This is jobID 2.
3. This makes our job table look like the follow. 

| jobID  | state | created | nextRun | retryAfter |
| -------| ----- | ------- | ------- | ---------- |
| 1 | QUEUED | 04:00 | 04:00 | +125 MINUTES |
| 2 | QUEUED | 05:00 | 05:00 | +125 MINUTES |

4. The agent loads a chat at 06:00. Given that we will fetch the earliest `nextRun` that is after the current time, we will load `jobID` 1, and update its `nextRun` and `state`. `nextRun` is set to the current time with the `retryAfter` modifier. Our table now looks like this 

| jobID  | state | created | nextRun | retryAfter |
| -------| ----- | ------- | ------- | ---------- |
| 1 | RUNQUEUED | 04:00 | 08:05 | +125 MINUTES |
| 2 | QUEUED | 05:00 | 05:00 | +125 MINUTES |

5. At 07:00, the agent exits the chat cleanly without responding, and we requeue the chat.

# Current implementation
6. When updating the job from the requeue command, we will set the `nextRun` time to current time, leaving our table like this.

| jobID  | state | created | nextRun | retryAfter |
| -------| ----- | ------- | ------- | ---------- |
| 1 | QUEUED | 04:00 | 07:00 | +125 MINUTES |
| 2 | QUEUED | 05:00 | 05:00 | +125 MINUTES |

7. When we fetch the next chat, we will follow the same logic as step 4, we will fetch `jobID` 2, which is incorrect, since it has been waiting less time.

# Your proposed implementation
6. When updating the job from the requeue command, we will set the `nextRun` time to the `nextRun` time with the inverted `retryAfter` modifier applied, leaving our table like this.

| jobID  | state | created | nextRun | retryAfter |
| -------| ----- | ------- | ------- | ---------- |
| 1 | QUEUED | 04:00 | 06:00 | +125 MINUTES |
| 2 | QUEUED | 05:00 | 05:00 | +125 MINUTES |

7. When we fetch the next chat, we will follow the same logic as step 4, we will fetch `jobID` 2, which is incorrect, since it has been waiting less time.

# My proposed implementation
6. When updating the job from the requeue command, we will set the `nextRun` time to the `created`, leaving our table like this.

| jobID  | state | created | nextRun | retryAfter |
| -------| ----- | ------- | ------- | ---------- |
| 1 | QUEUED | 04:00 | 04:00 | +125 MINUTES |
| 2 | QUEUED | 05:00 | 05:00 | +125 MINUTES |

7. When we fetch the next chat, we will follow the same logic as step 4, we will fetch `jobID` 1, which is correct, since it has been waiting the most time.


# Other things to note

- Only unanswered inputs have jobs, so any closed chat will not have a bedrock job.

Reply:
ok, that makes sense, thanks for the explanation.